pxtlib/auth.ts

```diff
@@ -305,59 +305,92 @@ namespace pxt.auth {
-
-        public async patchUserPreferencesAsync(ops: ts.pxtc.jsonPatch.PatchOperation | ts.pxtc.jsonPatch.PatchOperation[], immediate = false) {
-            ops = Array.isArray(ops) ? ops : [ops];
-            ops = ops.filter(op => !!op);
-            if (!ops.length) { return; }
-            const curPref = await this.userPreferencesAsync();
-            ts.pxtc.jsonPatch.patchInPlace(curPref, ops);
-            await this.setUserPreferencesAsync(curPref);
-            // If we're not logged in, non-persistent local state is all we'll use
-            if (!await this.loggedInAsync()) { return; }
+        private patchQueue: {
+            ops: ts.pxtc.jsonPatch.PatchOperation[];
+            filter?: (op: ts.pxtc.jsonPatch.PatchOperation) => boolean;
+        }[] = [];
+
+        public async patchUserPreferencesAsync(patchOps: ts.pxtc.jsonPatch.PatchOperation | ts.pxtc.jsonPatch.PatchOperation[], opts: {
+            immediate?: boolean,                                            // Sync the change with cloud immediately, don't debounce
+            filter?: (op: ts.pxtc.jsonPatch.PatchOperation) => boolean      // Filter the final set of patch operations
+        } = {}) {
+            patchOps = Array.isArray(patchOps) ? patchOps : [patchOps];
+            patchOps = patchOps.filter(op => !!op);
+            if (!patchOps.length) { return; }
+
+            const patchDiff = (pSrc: UserPreferences, ops: ts.pxtc.jsonPatch.PatchOperation[], filter?: (op: ts.pxtc.jsonPatch.PatchOperation) => boolean) => {
+                // Apply patches to pDst and return the diff as a set of new patch ops.
+                const pDst = U.deepCopy(pSrc);
+                ts.pxtc.jsonPatch.patchInPlace(pDst, ops);
+                let diff = ts.pxtc.jsonPatch.diff(pSrc, pDst);
+                // Run caller-provided filter
+                if (diff.length && filter) { diff = diff.filter(filter); }
+                return diff;
+            }
 
-            // If the user is logged in, save to cloud, but debounce the api call as this can be called frequently from skillmaps
-
-            // Replace matching patches in the queue
-            ops.forEach((incoming, iIncoming) => {
-                this.prefPatchOps.some((existing, iExisting) => {
-                    if (!ts.pxtc.jsonPatch.opsAreEqual(existing, incoming)) return false;
-                    // Patches are equivalent, replace in queue
-                    this.prefPatchOps[iExisting] = incoming;
-                    // Clear from incoming so we don't add it below
-                    (ops as ts.pxtc.jsonPatch.PatchOperation[])[iIncoming] = null;
-                    return true;
-                });
-            });
-            // Add remaining ops to the queue
-            ops.filter(op => !!op).forEach(op => this.prefPatchOps.push(op));
+            // Process incoming patch operations to produce a more fine-grained set of diffs. Incoming patches may be overly destructive
+            {
```

Comment:
this is just a bit nitpicky but is there a reason for the closure here / concern of variables conflicting or is it just for visually separating it out within the function? We don't tend to use non-behavioral closures too often so it stood out a bit (I don't know if there's an official term for this, I just mean 'a closure that only separates scope and does not have a keyword affecting it').

Reply:
Just scoping the variables. I use `diff` and `curPrefs` a lot in this function. But @livcheerful also commented on it being confusing. I'll clean it up :).